db/fetcher/indexer_iterators.go

```diff
@@ -423,70 +401,22 @@ func createIndexIterator(
 			return nil, err
 		}
 
+		m := &indexByteValuesMatcher{value: valueBytes}
 		switch op {
 		case opEq:
-			if isUnique {
-				return &eqSingleIndexIterator{
-					indexKey: indexDataStoreKey,
-					filterValueHolder: filterValueHolder{
-						value: valueBytes,
-					},
-					execInfo: execInfo,
-				}, nil
-			} else {
-				return &eqPrefixIndexIterator{
-					indexKey: indexDataStoreKey,
-					filterValueHolder: filterValueHolder{
-						value: valueBytes,
-					},
-					execInfo: execInfo,
-				}, nil
-			}
+			m.evalFunc = func(res int) bool { return res == 0 }
```

Comment:
question: As mentioned, I really do like the improvements to this switch, however it does feel a bit strange that you are returning an interface and instead of having lots of implementations for that interface, you are passing in a func to a private property in a concrete implementation of that interface and then returning that one 'super' type instance.

Why did you chose to do this?  Did it just kind of evolve as you were working in this space?

Reply:
the interface is different kinds of matchers there is byte-matcher, in-matcher, like-matcher and new any-matcher. All of them (except for any) are configurable. Byte-matcher just a bit more: it compares bytes and we can configure what is the expected result of the comparison. 

I think this way it's a bit more concise, but I don't have a strong preference here.

Reply:
Ah okay, there are still a fair few concrete types then :)

> I think this way it's a bit more concise, but I don't have a strong preference here.

Keep it as-is then, unless anyone else has a really strong preference - I have nothing really against this, it just struck me as unusual and worth highlighting/chatting about :)  Is a fairly nice way of reducing the number to types required.